cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/cask360/Cask360TableDataset.java

```diff
@@ -0,0 +1,424 @@
+   * be updated with the specified value.
+   * <p>
+   * This operation currently never fails for data modeling reasons and performs
+   * no type or consistency checks across groups or entities.
+   *
+   * @param entity
+   */
+  public void write(Cask360Entity entity) {
+    write(entity.getID(), entity);
+  }
+
+  /**
+   * Writes the specified {@link Cask360Entity} to the table. Additional API
+   * call to support the {@link BatchWritable} interface.
+   * <p>
+   * See {@link #write(Cask360Entity)} for more info on the semantics of this
+   * operation.
+   */
+  @Override
+  public void write(String id, Cask360Entity entity) {
+    List<byte[][][]> dataArrays = new LinkedList<byte[][][]>();
+    for (Map.Entry<String, Cask360Group> entry : entity.getGroups().entrySet()) {
+      Cask360Group group = entry.getValue();
+      Cask360GroupMeta groupMeta = getGroupMeta(Bytes.toBytes(entry.getKey()), group.getType());
+      if (group.getType() != groupMeta.getType()) {
+        String msg = "Group type mismatch, attempted (" + group.getType() + ") existing (" + groupMeta.getType() + ")";
+        LOG.warn(msg);
+        throw new IllegalArgumentException(msg);
+      }
+
+      dataArrays.add(Cask360TableDataset.mapToArrays(group.getData().getBytesMap(groupMeta.getPrefix())));
+    }
+    byte[][][] dataArray = Cask360TableDataset.flattenDataArrays(dataArrays);
+    this.dataTable.put(Bytes.toBytes(id), dataArray[0], dataArray[1]);
+  }
+
+  /**
+   * Reads the {@link Cask360Entity} for the specified ID.
+   * <p>
+   * If no entity exists for the specified ID, returns null.
+   *
+   * @param id
+   *          entity ID to read
+   * @return entity data, or null if entity has no data
+   */
+  public Cask360Entity read(String id) {
+    Row row = this.dataTable.get(Bytes.toBytes(id));
+    if ((row == null) || row.isEmpty()) {
+      return null;
+    }
+    return new Cask360Entity(id, rowToGroupMap(row));
+  }
+
+  // Group Management Internal Methods
+
+  static class Cask360GroupCache {
+
+    /** Local cache mapping group names to group numbers. */
+    private Map<byte[], Cask360GroupMeta> groupPrefixes = null;
+
+    /** Local cache mapping group numbers to group names. */
+    private Map<byte[], Cask360GroupMeta> groupPrefixesRev = null;
+
+    Cask360GroupCache() {
+      this.groupPrefixes = new TreeMap<byte[], Cask360GroupMeta>(Bytes.BYTES_COMPARATOR);
+      this.groupPrefixesRev = new TreeMap<byte[], Cask360GroupMeta>(Bytes.BYTES_COMPARATOR);
+    }
+
+    Cask360GroupMeta get(byte[] name) {
+      return this.groupPrefixes.get(name);
+    }
+
+    Cask360GroupMeta get(short number) {
+      return this.groupPrefixesRev.get(Cask360GroupMeta.getPrefix(number));
+    }
+
+    void put(Cask360GroupMeta meta) {
+      this.groupPrefixes.put(meta.getName(), meta);
+      this.groupPrefixesRev.put(meta.getPrefix(), meta);
+    }
+  }
+
+  private synchronized Cask360GroupMeta getGroupMeta(byte[] name, Cask360GroupType type) {
```

Comment:
There is a local cache in the instance of the dataset. The Cask360Table has a meta table that stores the different groups, their ids and configuration. There is a local write-through cache in each instance of the dataset to store this data, otherwise you will have to do a lookup into the meta table for read/write every operation.

If you happen to use the same instance in multiple threads, this will protect you. Since datasets aren't threadsafe, should I purposefully make this not threadsafe?